src/middlewared/middlewared/plugins/mail.py

```diff
@@ -1,41 +1,324 @@
+            while True:
+                read = f.read(1048576)  # 1MiB
+                if read == b'':
+                    break
+                data += read
+                i += 1
+                if i > 50:
+                    raise ValueError('Attachments bigger than 50MB not allowed yet')
+            if data == b'':
+                return None
+            return json.loads(data)
+
+        attachments = read_json() if message.get('attachments') else None
+        if attachments:
+            msg = MIMEMultipart()
+            msg.preamble = message['text']
+            for attachment in attachments:
+                m = Message()
+                m.set_payload(attachment['content'])
+                for header in attachment.get('headers'):
+                    m.add_header(header['name'], header['value'], **(header.get('params') or {}))
+                msg.attach(m)
+        else:
+            msg = MIMEText(message['text'], _charset='utf-8')
+
+        subject = message.get('subject')
+        if subject:
+            msg['Subject'] = subject
+
+        msg['From'] = config['fromemail']
+        msg['To'] = ', '.join(to)
+        msg['Date'] = formatdate()
+
+        local_hostname = socket.gethostname()
+
+        msg['Message-ID'] = "<%s-%s.%s@%s>" % (sw_name.lower(), datetime.utcnow().strftime("%Y%m%d.%H%M%S.%f"), base64.urlsafe_b64encode(os.urandom(3)), local_hostname)
+
+        extra_headers = message.get('extra_headers') or {}
+        for key, val in list(extra_headers.items()):
+            if key in msg:
+                msg.replace_header(key, val)
+            else:
+                msg[key] = val
+
+        try:
+            server = self._get_smtp_server(config, message['timeout'], local_hostname=local_hostname)
+            # NOTE: Don't do this.
+            #
+            # If smtplib.SMTP* tells you to run connect() first, it's because the
+            # mailserver it tried connecting to via the outgoing server argument
+            # was unreachable and it tried to connect to 'localhost' and barfed.
+            # This is because FreeNAS doesn't run a full MTA.
+            # else:
+            #    server.connect()
+            syslog.syslog("sending mail to " + ','.join(to) + msg.as_string()[0:140])
+            server.sendmail(config['fromemail'], to, msg.as_string())
+            server.quit()
+        except ValueError as ve:
+            # Don't spam syslog with these messages. They should only end up in the
+            # test-email pane.
+            raise CallError(str(ve))
+        except Exception as e:
+            self.logger.warn('Failed to send email: %s', str(e), exc_info=True)
+            if message['queue']:
+                with MailQueue() as mq:
+                    mq.append(msg)
+            raise CallError(f'Failed to send email: {e}')
+        except smtplib.SMTPAuthenticationError as e:
+            raise CallError(f'Authentication error ({e.smtp_code}): {e.smtp_error}', errno.EAUTH)
```

Comment:
Why `SMTPAuthenticationError` is different from any other SMTP server misconfiguration? (Failed messages are not retried)

Reply:
I dont get what you're saying here. We dont want authentication errors to be queued? Can you explain a little further?

Reply:
Typo in server hostname and typo in password are both configuration errors, but we retry messages when first error occurs (in `except Exception`) and discard them when second error occurs (in `smtplib.SMTPAuthenticationError`). This seems illogical to me.

Also, I've just noticed, `except smtplib.SMTPAuthenticationError` is located after `except Exception`, so it will not be executed ever and all messages will be retried. If this is desired behavior, we can just delete lines 277 and 278.

Reply:
For the exact reason it is a configuration error the message should not be retried. Same codepath is also used to test send mail.

You're right about SMTPAuthenticationError being after Exception, it was a copy+paste, will fix that.